[docs] Coin registry (#23750)
MystenLabs/sui: docs/content/guides/developer/currency.mdx

---
title: Create Currencies and Tokens

> description: Learn how to create currencies and mint coins and tokens on the Sui network using legacy coin creation or the new Coin Registry system.

Reviewer: ```suggestion
description: Learn how to create and mint currencies on the Sui network using legacy coin creation or the new Coin Registry system.
```

keywords: [ mint tokens, create tokens, use tokens, publish token, custom tokens, sui CLI, denylist, regulated coins, coin registry, currency registry ]
pagination_prev: null
---

The Coin Registry system provides a centralized approach to currency management through the `sui::coin_registry` module. The registry is a shared object located at address `0xc` that stores metadata, supply information, and regulatory status for all registered coin types.

> The `sui::token` module handles token creation on the network. Refer to the [Closed-Loop Token](../../standards/closed-loop-token.mdx) and [Currency](../../standards/currency.mdx) Standards documentation for more information on these features.

Reviewer: ```suggestion
The `sui::token` module handles token creation on the network. Refer to the [Closed-Loop Token](/standards/closed-loop-token.mdx) and [Currency](/standards/currency.mdx) Standards documentation for more information on these features.
```

### Currency creation process

The registry system supports two currency creation methods:

> 1. **Standard creation:** Call `new_currency<T>()` when creating the coin outside of the init function of your package.

Reviewer: ```suggestion
1. **Standard creation:** Call `new_currency<T>()` when creating the coin outside of the `init` function of your package.
```

> 2. **OTW creation:** Use `new_currency_with_otw<T>()` with a One-Time Witness for uniqueness proof.

Reviewer: ```suggestion
1. **OTW creation:** Use `new_currency_with_otw<T>()` with a One-Time Witness for uniqueness proof.
```

Both methods return a `CurrencyInitializer<T>` that allows configuration before finalization:

<details>
<summary>
Regular currency creation
</summary>
<ImportContent source="examples/move/coin/sources/non_otw_currency.move" mode="code" />
</details>

<details>
<summary>
OTW currency creation
</summary>
<ImportContent source="examples/move/coin/sources/my_coin_new.move" mode="code" />
</details>

The initialization process allows for:

> - **Supply model selection:** Choose fixed, burn-only, or flexible supply
> - **Regulatory configuration:** Add deny list capabilities if needed

Reviewer: ```suggestion
- **Supply model selection:** Choose fixed, burn-only, or flexible supply.
- **Regulatory configuration:** Add deny list capabilities if needed.
```

:::caution Important

After initialization of a currency using the OTW method, you must call `finalize_registration` to create the shared `Currency` object that the Coin Registry can track.

:::

## `DenyList`

> The Sui framework provides a `DenyList` singleton, shared object that the bearer of a `DenyCapV2` can access to specify a list of addresses that are unable to use a Sui core type. The initial use case for `DenyList`, however, focuses on limiting access to coins of a specified type. This is useful, for example, when creating a regulated coin on Sui that requires the ability to block certain addresses from using it as inputs to transactions. Regulated coins on Sui satisfy any regulations that require the ability to prevent known bad actors from having access to those coins. 

Reviewer: ```suggestion
The Sui framework provides a `DenyList` shared object that the bearer of a `DenyCapV2` can access to specify a list of addresses that are unable to use a Sui type. The initial use case for `DenyList`, however, focuses on limiting access to coins of a specified type. This is useful, for example, when creating a regulated coin on Sui that requires the ability to block certain addresses from using it as inputs to transactions. Regulated coins on Sui must satisfy any regulations that require the ability to prevent known bad actors from having access to those coins. 
```
